include/pybind11/cast.h

```diff
@@ -34,6 +34,20 @@ template <typename Type> class opaque {
 
 NAMESPACE_BEGIN(detail)
 
+/// Convert from any type pointer to another avoding warnings about void* vs (void*()) incompatibility.
```

Comment:
Yep, looks like it might be it.

Anyhow, I'll just use patched version of pybind11 for now, and will see how future versions of gcc behave. Who knows maybe one day we'll get modules for c++ and this kind of problem will go away forever.

Thanks for looking into it. We can probably close this now.